### ISSUE TITLE
ImageNet?

### ISSUE THREAD

**Issue by tjingrant** [hongyi-zhang/Fixup#1]
Hi, thanks for the implementation! I'd love to try out the imagenet version of Fixup Resnet; I noticed you wrote this in the README:

> A re-implementation is available. However, I have not been able to test it. If you try it out, please feel free to contact me --- your feedback is very welcome!

Anything I can help? What's the obstacle? I'm very interested in seeing the results on imagenet and would like to help in any ways I can.

Again thanks for the wonder work!

**Comment by hongyi-zhang** [comment]
Hi, thanks for your comments and offering to help! The experiments reported in the paper was performed using Facebook's computing resource. Now that all the authors have left Facebook, we no longer have access to the code and computing resources.

If you have PyTorch, GPUs and ImageNet dataset ready, it should be very easy to try out the implementation. Thanks for your help!

**Comment by tjingrant** [comment]
@hongyi-zhang got it, thanks. Will start a run either today or tomorrow.

**Comment by tjingrant** [comment]
@hongyi-zhang started a run by importing your fixup models in the torch imagenet example wrapper. Used the same set of hyper parameters as official torch ResNet:
```
 python main.py -a fixup_resnet50 [data dir]
```

Based on experience, it takes ~3 days to finish and so I'll get back to you in 3 days. After that, I can probably help run ResNet18 as well.  BTW, do you think the hyper parameters needs any modifications/changes compared to the ones used to train normal ResNet?

**Comment by hongyi-zhang** [comment]
@tjingrant Thank you for your help!

As noted in our paper, ERM training works with the default hyper parameters, but the test performance will not be good. If using Mixup (e.g., alpha=0.7) we recommend lowering the learning rate (e.g. to 0.1x) for all the scalar parameters, as they tend to have large gradients in the beginning epochs.

Please let me know once you get the results. Thanks :)

**Comment by tjingrant** [comment]
@hongyi-zhang thanks for the heads-up; I did not modify the learning rates and this could explain why it's having trouble converging. I saw your code to reduce learning rates for scalar parameters, will adopt it and restart the training tmrw.

**Comment by tjingrant** [comment]
I'm having similar experience with https://github.com/hongyi-zhang/Fixup/issues/2.
I'm still having trouble converging even after applying your suggestions.
I've uploaded my training code to Github and you can see the specific edit I did to the usual imagenet driver code: https://github.com/tjingrant/fixup-driver/commit/90bd25f187d1f99a8e062262438340ebd1eee827 .

I can upload my training log as well if there's such a need.

**Comment by hongyi-zhang** [comment]
I think in your code the scalar parameters still have the default high learning rate due to this line which sets each parameter group to the same global learning rate: https://github.com/tjingrant/fixup-driver/blob/90bd25f187d1f99a8e062262438340ebd1eee827/imagenet/main.py#L393-L394

In the CIFAR-10 trainer, we rescale the current global learning rate based on the initial learning rate of each parameter group: https://github.com/hongyi-zhang/Fixup/blob/4351e4d939654d9b8a5ceab20b43005055b0c0e2/cifar/cifar_train.py#L210-L218

(The 'initial_lr' property is created by calling a learning rate scheduler constructor.)

**Comment by tjingrant** [comment]
thanks @hongyi-zhang! I adjusted my driver code according to your suggestions today. Trained for a few epochs and convergence looks great so far! Will get back to you in 3 days!

**Comment by tjingrant** [comment]
@hongyi-zhang the best accuracy @ 1 i got is 75.556 for resnet50, is this good enough?
Do you want me to upload my code via a pull request?

**Comment by hongyi-zhang** [comment]
@tjingrant that’s a reasonably good number, but we certainly got a bit (0.4-0.5) higher; some of the difference is likely due to a few implementation details.
Did you run 100 epochs with a final learning rate drop at 90th epoch?

Yes, please make a pull request and I will merge it with the master. Thanks!

**Comment by tjingrant** [comment]
@hongyi-zhang I only ran 90 epochs.

**Comment by tjingrant** [comment]
I could certainly try to get another 10 epochs run.

**Comment by hongyi-zhang** [comment]
Yes, please try to add another 10 epochs with 0.001x maximal learning rate -- it could be the reason.

**Comment by hongyi-zhang** [comment]
@tjingrant Any update on the 100 epoch experiments?

**Comment by tjingrant** [comment]
@hongyi-zhang really sorry I got some research ddls coming up, will get back to you within 2-3 days.

**Comment by tjingrant** [comment]
@hongyi-zhang do you want me to commit the code now since it's more or less correct?

**Comment by hongyi-zhang** [comment]
@tjingrant Yes, you can commit the code at any time. Thanks!

**Comment by tjingrant** [comment]
@hongyi-zhang so I re-ran with the last 10 epochs and best accuracy @ 1 is 
```
 * Acc@1 76.012
```
I've submitted a pull request, please help review, thanks!

**Comment by hongyi-zhang** [comment]
Great, thanks! It looks good to me.

**Comment by blueardour** [comment]
Hi, @hongyi-zhang Inspiring work. Also great thanks for the sharing of the code.

@tjingrant  @hongyi-zhang I wonder if ever the Resnet-18 was tried.  I'm reproducing the result of 'fixup initilization', however, with my limited GPU resources, it takes me too long time to train the Resnet50. Maybe Resnet18 is more suitable for quick evaluation?

**Comment by lopuhin** [comment]
Hi @hongyi-zhang @tjingrant very inspiring work, thank you for taking care of the code and reproducibility! 

I wonder if anyone has imagenet weights? I wanted to check transfer learning perfrormance.
